assets/src/stories-editor/components/higher-order/with-replaced-full-size-image.js

```diff
@@ -0,0 +1,66 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { dispatch, select } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { MAX_IMAGE_SIZE_SLUG } from '../../constants';
+
+/**
+ * A component that extends the BlockEdit components, replacing the 'full' image size with a custom max size.
+ *
+ * @param {Function} BlockEdit A block's edit component to be extended.
+ * @return {Function} A function returning an extended BlockEdit.
+ */
+export default ( BlockEdit ) => {
+	return class ReplacedFullSizeImage extends BlockEdit {
+		/**
+		 * The class constructor.
+		 */
+		constructor() {
+			super( ...arguments );
+			this.componentDidUpdate = this.componentDidUpdate.bind( this );
+			this.ampReplaceFullSizeImage = this.ampReplaceFullSizeImage.bind( this );
+		}
+
+		/**
+		 * When the component updates, if it's for the Image block, replace the full size image.
+		 */
+		componentDidUpdate() {
+			const { name } = this.props;
+			if ( 'core/image' === name ) {
+				this.ampReplaceFullSizeImage();
+			}
+
+			if ( super.componentDidUpdate ) {
+				super.componentDidUpdate();
+			}
+		}
+
+		/**
+		 * Replaces the 'full' image size with a custom size, which has a limited height.
+		 *
+		 * This prevents the user from selecting an image in the Image block that's too big.
+		 * For example, 3200 x 4000.
+		 */
+		ampReplaceFullSizeImage() {
+			const blockEditorStore = 'core/block-editor';
+			const initialSizes = select( blockEditorStore ).getSettings( 'imageSizes' ).imageSizes;
+			const sizesWithoutFullSize = initialSizes.filter( ( size ) => 'full' !== size.slug );
+
+			// If the AMP Story slug isn't present, add it.
+			if ( ! sizesWithoutFullSize.filter( ( size ) => MAX_IMAGE_SIZE_SLUG === size.slug ).length ) {
```

Comment:
It'd be nice if this logic to add the image size didn't have to run so often. For example, if it ran only on `domready`.

But it looks like the settings update, so these settings will be overwritten if this logic only runs on `domready`.